### PR TITLE
dstore: Fix extension slot allocation

### PR DESCRIPTION
**Pull request by artpol84**
Signed-off-by: Artem Polyakov <artpol84@gmail.com>

### CODE REVIEW

**Comment by artpol84** [comment]
Address https://github.com/pmix/master/issues/280
Need to verify that it fixes it.

**Comment by artpol84** [comment]
Jenkins failed because https://github.com/pmix/master/pull/278 is not merged in the master yet (as I noted on today's telecon).
#283 went well because v1.2 is free from this bug.

**Comment by artpol84** [comment]
@rhc54  this is ok to merge.

**Comment by artpol84** [comment]
@hjelmn was going to test this tonight.what is the right way to provide him with this changes?

**Comment by hjelmn** [comment]
I just took the patch from your repo and applied it to my branch. I wasn't able to test last night though. Will try again today/tonight.

**Comment by hjelmn** [comment]
:+1: looks like this fixes the issue.

**Comment by artpol84** [comment]
@hjelmn great news! Thank you.

**Comment by hjelmn** [comment]
Hmm, seeing this now with some runs. Not sure if it is related:

```
[nid20703:83300] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2201
[nid20703:83300] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2396
[nid20703:83300] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 1094
[nid20703:83300] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 97
[nid20703:83300] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 376
[nid20703:83300] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/server/pmix_server.c at line 463
[nid20761:83632] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2201
[nid20761:83632] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2396
[nid20761:83632] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 1094
[nid20761:83632] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 97
[nid20761:83632] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 376
[nid20761:83632] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/server/pmix_server.c at line 463
```

**Comment by artpol84** [comment]
you fully applied my patch? Can I rely on those line numbers?
can you maybe provide links in github to corresponding lines?
what is the scale?

**Comment by artpol84** [comment]
Looks like line numbers are OK

**Comment by artpol84** [comment]
This is something else (in case error lines are correct, and it's seems to be so).
[This error](https://github.com/artpol84/master/blob/master/src/dstore/pmix_esh.c#L2201) happens when `put_data_to_the_end` returned error. The thing that puzzles me is that this function should also error-output on each error exit:
* https://github.com/artpol84/master/blob/master/src/dstore/pmix_esh.c#L2133
* https://github.com/artpol84/master/blob/master/src/dstore/pmix_esh.c#L2142 
* https://github.com/artpol84/master/blob/master/src/dstore/pmix_esh.c#L2150 (here btw it should return 0 instead of ERROR, I've fixed that in additional commit)

Can you double check your output for one of those messages?

**Comment by artpol84** [comment]
Also can you try with my new updated PR. 
Returning `PMIX_ERROR` which is negative [here](https://github.com/artpol84/master/blob/master/src/dstore/pmix_esh.c#L2151) in the `put_data_to_the_end` which returns `size_t` is incorrect and may theoretically cause a problem.

**Comment by artpol84** [comment]
By the way! There is already a way to set the size of a segment, no need to implement this.
provide `ESH_ENV_NS_DATA_SEG_SIZE` for both `orted` and application processes.

**Comment by artpol84** [comment]
Just in case - could you check that the storage of the dstore files is not overflowed.

**Comment by hjelmn** [comment]
That run was 512 nodes @ 68ppn. I didn't see the error until I added --mca orte_debug_daemons true .

**Comment by hjelmn** [comment]
BTW, the program I was running was hostname.

**Comment by hjelmn** [comment]
With both patches:

```
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2201
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2396
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 1094
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 97
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 376
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/server/pmix_server.c at line 463
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2201
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2396
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 1094
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 97
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 376
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/server/pmix_server.c at line 463
```

**Comment by artpol84** [comment]
Lets try the following patch to debug:

```patch
diff --git a/src/dstore/pmix_esh.c b/src/dstore/pmix_esh.c
index da465d6..5a1eb45 100644
--- a/src/dstore/pmix_esh.c
+++ b/src/dstore/pmix_esh.c
@@ -2129,6 +2129,15 @@ static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg,
     if (sizeof(size_t) + ESH_KEY_SIZE(key, size) + EXT_SLOT_SIZE() > _data_segment_size) {
         /* this is an error case: segment is so small that cannot place evem a single key-value pair.
          * warn a user about it and fail. */
+
+        PMIX_ERROR_LOG(PMIX_ERR_NOMEM);
+{
+    int delay = 1;
+    while( delay ){
+        sleep(1);
+    }
+}
+
         offset = 0; /* offset cannot be 0 in normal case, so we use this value to indicate a problem. */
         pmix_output(0, "PLEASE set NS_DATA_SEG_SIZE to value which is larger when %lu.",
 		sizeof(size_t) + strlen(key) + 1 + sizeof(size_t) + size + EXT_SLOT_SIZE());
@@ -2140,6 +2149,14 @@ static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg,
         tmp = extend_segment(tmp, &ns_info->ns_map);
         if (NULL == tmp) {
             PMIX_ERROR_LOG(PMIX_ERR_NOMEM);
+
+{
+    int delay = 1;
+    while( delay ){
+        sleep(1);
+    }
+}
+
             offset = 0; /* offset cannot be 0 in normal case, so we use this value to indicate a problem. */
             return offset;
         }
@@ -2148,6 +2165,14 @@ static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg,
         ns_seg_info_t *elem = _get_ns_info_from_initial_segment(&ns_info->ns_map);
         if (NULL == elem) {
             PMIX_ERROR_LOG(PMIX_ERR_NOMEM);
+
+{
+    int delay = 1;
+    while( delay ){
+        sleep(1);
+    }
+}
+
             offset = 0; /* offset cannot be 0 in normal case, so we use this value to indicate a problem. */
             return offset;
         }
@@ -2155,6 +2180,15 @@ static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg,
         offset = sizeof(size_t);
     }
     global_offset = offset + id * _data_segment_size;
+
+if( 0 == global_offset) {
+    int delay = 1;
+    while( delay ){
+        sleep(1);
+    }
+}
+
+
     addr = (uint8_t*)(tmp->seg_info.seg_base_addr)+offset;
     ESH_PUT_KEY(addr, key, buffer, size);
 ```

**Comment by artpol84** [comment]
@jjhursey I guess that @hjelmn stated that the original issue he was observing went away with this fix.
There might be additional problem there - we are debugging now. But this fix we need 100%.

**Comment by hjelmn** [comment]
The other issue I observed is not re-appearing. Will open a new bug if it shows up again.

:+1:

File: src/dstore/pmix_esh.c [hunk 1]
```diff
@@ -2091,15 +2094,15 @@ static int put_empty_ext_slot(seg_desc_t *dataseg)
     uint8_t *addr;
     global_offset = get_free_offset(dataseg);
     rel_offset = global_offset % _data_segment_size;
-    if (rel_offset + EXT_SLOT_SIZE(ESH_REGION_EXTENSION) > _data_segment_size) {
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: Need to check if we need to fix `0 == rel_offset` case here, will check tomorrow.

**Pull request merged**